MessageSigning/MessageSigning.md

```diff
@@ -0,0 +1,433 @@
+  body_protected : empty_or_serialized_map,                     ; protected from layer 1
+  ? sign_protected : empty_or_serialized_map,                   ; not present in Sign1 case
+  external_aad : bstr,                                          ; explanation below
+  payload : bstr
+]
+```
+
+The `external_aad` allows an application to ask the user to sign some extra data but NOT put it inside the COSE structure (only as part of the data to sign). Defaults to `h''`. You can read more about this at [RFC 8152 section-4.3](https://tools.ietf.org/html/rfc8152#section-4.3).
+
+The `context` is meant to encode what structure was used for the COSE request. `CounterSignature` is explained in a later section of this specification.
+
+#### Signing and Verification process
+
+To be able to effectively verify we need two things:
+
+1) P1 - (optional) knowledge of the relation of a public key and a Cardano address
+1) P2 - Knowledge of which algorithm was used to sign
+
+For `P1`, just knowing the public key is not sufficient to know the address (which we need to know if we want to prove ownership of a UTXO or transaction). To resolve this, first we (optionally) add to the header an extra field.
+
+```
+? address: bstr,
+? address_pubkey: bstr,
+```
+
+Both `address` and `address_pkey` are optional so that if the receiver knows this information or knows the mapping between the public key and the address, they can exclude them to reduce the payload size.
+
+In the `v2` scheme, we need to specify the public key (since `address` only contains the `Blake2b-224` hash of the public key)
+The `v3` scheme is similar but there can be both a staking and spending credential in base addresses, but it is trivial to check which of the two corresponds to `address_pubkey`.
+As `v2` and `v3` addresses are incompatible, we can simply try both options until we reach the one that matches instead of requiring additional version data.
+
+For `P2`, the algorithm that is used is uniquely determined by the public key used. *COSE* defines the `alg` pre-defined header to indicate which algorithm was used to sign/verify but we don't use this for the following reasons
+
+1) We want to limit the number of allowed signing algorithms to exactly the set used in Cardano because of `E2`.
+1) There is a 1-1 mapping between the pair (`address`, `address_pubkey`) and what signing algorithm is used and so this information would be redundant.
+1) Allowed values in the `alg` field must be present in the [IANA COSE Algorithms registry](https://www.iana.org/assignments/cose/cose.xhtml#key-type) (and new entries have to be manually approved). This is not sufficient for us as we often have to use cutting edge cryptography specialized for our domain.
+
+### Encryption
+
+Although COSE defines multiple ways to encrypt, we simplify our spec to the two following cases:
+
+1) Encrypted with the recipient's public key (called `key transport` in COSE spec)
```

Comment:
How would that work in practice? I don't recall seeing any ed25519 library which allowed public-key encryption. They only typically allow private-key signing. So I wonder if it's even possible to do such thing with Ed25519? 

COSE speaks about Diffie-Hellman and RSA is also a possibility. I found a few things in the literature using ECDSA which could suggest that this is achievable for Ed25519 curves but that's definitely something to ask cryptographers about?

Reply:
How to handle ed25519 keys is explained in the header "Public key based encryption"

Reply:
Yup sorry, I read through that later on and saw how the document defines it with Diffie-Hellman shared secret, should've removed the comment :+1: